### PR TITLE
Fixed exam authorization command and refactoring

### PR DESCRIPTION
**Pull request by amir-qayyum-khan**
#### What are the relevant tickets?
- fixes https://github.com/mitodl/micromasters/issues/2445
- Fixed ripples of #2331

#### What's this PR do?
- Refactored the code of exam authorization command
- Added exceptions and logging to see behaviour of code
- Added new tests for authorization code.

### How to test:
- In .env add `FEATURE_FINAL_GRADE_ALGORITHM=v1`
- Set program.exam_series_code
- Set course.exam_module
- Pay for course
- Create final grade http://192.168.99.100:8079/admin/grades/finalgrade/
- Run command ```docker-compose run web ./manage.py authorization_user_exam --username=(value) --program-id=(value)```
- Check weather ( ExamAuthoriztion) object created 
  - ``docker-compose run web ./manage.py shell``
  - ``from exams.models import ExamAuthorization``
  -  ``ExamAuthorization.objects.all()``

#### To create dummy data:
```docker-compose run web ./manage.py shell```

```shell
from courses.models import CourseRun
from dashboard.models import CachedEnrollment
from dashboard.factories import CachedEnrollmentFactory
from ecommerce.factories import (OrderFactory, LineFactory)
from django.contrib.auth.models import User
from exams.models import ExamAuthorization

user = User.objects.get(username='amirq')
# use same  `edx_course_key` for final grade entry
course_run = CourseRun.objects.get(edx_course_key='course-v1:MITx+Digital+Learning+100+Jan_2016')
CachedEnrollmentFactory.create(user=user, course_run=course_run)
# enter  `total_price_paid` according to income threshold
order = OrderFactory.create(user=user, status='fulfilled', total_price_paid=1200)
LineFactory.create(order=order, course_key=course_run.edx_course_key)
ExamAuthorization.objects.all()
```

@pdpinch @giocalitri

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -230,6 +230,22 @@ You should now be able to do the following:
...
+ 
+ **params**
+ - **--username:** (optional) If provided, command will look for given user only, 
+    and authorize him if he matches the critaria.
+ - **--program-id:** (optional) if provided, command will look for given program only, 
+    and authorize all users who matches the critaria.
+ - **--course-key:**  (optional) if provided, command will look for given course only, 
+    and authorize all users who matches critaria.
+ 
+```python
```
Comment (rhysyngsun) [inline README.md]: This should be ` ```shell ` instead.

File: exams/signals.py [hunk 2]
```diff
@@ -29,10 +29,10 @@ def update_exam_authorization_final_grade(sender, instance, **kwargs):  # pylint
     authorize_for_exam(mmtrack, instance.course_run)
 
 
-@receiver(post_save, sender=CachedEnrollment, dispatch_uid="update_exam_authorization_cached_enrollment")
-def update_exam_authorization_cached_enrollment(sender, instance, **kwargs):  # pylint: disable=unused-argument
+@receiver(post_save, sender=CachedCurrentGrade, dispatch_uid="update_exam_authorization_grades")
```
Comment (rhysyngsun) [inline exams/signals.py]: Why is this changing to `CachedCurrentGrade`, could you give some context? My understanding was that `CachedCurrentGrade` was the non-final grade in our system so tt seems we lose the behavior of doing this when the user enrolls.

File: exams/utils.py [hunk 3]
```diff
@@ -59,3 +67,53 @@ def authorize_for_exam(mmtrack, course_run):
...
+                course_run.edx_course_key
+            )
+
+
+def bulk_authorize_for_exam(program_id=None, username=None, course_key=None):
+    """
+    Authorize user(s) for exam(s).
+
+    Args:
+        program_id(str): Program id (optional) if you want authorization on spcific program exams
```
Comment (rhysyngsun) [inline exams/utils.py]: Typo: spcific

File: exams/utils.py [hunk 4]
```diff
@@ -59,3 +67,53 @@ def authorize_for_exam(mmtrack, course_run):
...
+        return
+
+    for program in programs:
+        program_enrollment_qset = ProgramEnrollment.objects.filter(program=program)
+        if username:
+            program_enrollment_qset = program_enrollment_qset.filter(user__username=username)
+
+        if not program_enrollment_qset.exists():
+            log.info('[authorization_user_exam] User(s) not exist for exam authorization.')
+            return
```
Comment (rhysyngsun) [inline exams/utils.py]: Shouldn't this be a `continue`?

File: exams/utils.py [hunk 5]
```diff
@@ -59,3 +67,53 @@ def authorize_for_exam(mmtrack, course_run):
...
+
+        for program_enrollment in program_enrollment_qset:
+            user = program_enrollment.user
+            enrollments_qset = CachedEnrollment.user_course_qset(user, program=program)
+            if course_key:
+                enrollments_qset = enrollments_qset.filter(course_run__edx_course_key=course_key)
+
+            if not enrollments_qset.exists():
+                log.info('[authorization_user_exam] Course(s) not exist for exam authorization.')
+                return
```
Comment (rhysyngsun) [inline exams/utils.py]: Shouldn't this be a `continue`?

File: exams/utils.py [hunk 6]
```diff
@@ -59,3 +67,53 @@ def authorize_for_exam(mmtrack, course_run):
...
+def bulk_authorize_for_exam(program_id=None, username=None, course_key=None):
+    """
+    Authorize user(s) for exam(s).
+
+    Args:
+        program_id(str): Program id (optional) if you want authorization on spcific program exams
+        username(str): User name (optional) whom you want to authorize for exam(s)
+        course_key(str): Edx course key  (optional) where you want to authorize user(s)
+    """
+    programs = Program.objects.filter(exam_series_code__isnull=False, live=True).filter(~Q(exam_series_code=""))
```
Comment (rhysyngsun) [inline exams/utils.py]: I would suggest refactoring these filters a little so it's a little easier to read:

```python
Program.objects.filter(live=True).exclude(exam_series_code__isnull=True).exclude(exam_series_code__exact="")
```

File: exams/utils.py [hunk 7]
```diff
@@ -59,3 +67,53 @@ def authorize_for_exam(mmtrack, course_run):
...
+            if not enrollments_qset.exists():
+                log.info('[authorization_user_exam] Course(s) not exist for exam authorization.')
+                return
+
+            mmtrack = get_mmtrack(user, program)
+
+            for enrollment in enrollments_qset:
+                # if user has passed and paid for the course
+                # and not already authorized for exam the create authorizations.
+                authorize_for_exam(mmtrack, enrollment.course_run)
```
Comment (rhysyngsun) [inline exams/utils.py]: I'm not sure we want to authorize the user for *every* course_run the user ever took. @pdpinch I know you mentioned we want to check all the course runs, but do we want to do authorize for all runs or do something more targeted like only authorize for the most recent course run the user is eligible to take an exam for?

File: financialaid/api_test.py [hunk 8]
```diff
@@ -36,7 +36,7 @@
 from search.base import MockedESTestCase
 
 
-def create_program(create_tiers=True, past=False):
+def create_program(create_tiers=True, past=False, exam_series_code="", exam_module=""):
```
Comment (rhysyngsun) [inline financialaid/api_test.py]: I think `None` might be a better default here

File: exams/signals.py (hunk 2 above)
Comment (amir-qayyum-khan) [inline exams/signals.py]: actually we determine if user is pass or not using a process that involves current grade 
https://github.com/mitodl/micromasters/blob/master/dashboard/utils.py#L223-L233

I think i should also add a signal for cached certificates as well, because for not financial_aid programs

Comment (rhysyngsun) [inline exams/signals.py]: Are we fixing all these signals in the same PR? I'd advocate for addressing any changes to signals separately so this review doesn't drag out.

Comment (amir-qayyum-khan) [inline exams/signals.py]: i changes the title of PR and description is updated. I think we should fix every issue on exam authorization triggering here. To avoid complexity because code is already merge and think it has some issues in both signals and command

Comment (rhysyngsun) [inline exams/signals.py]: @amir-qayyum-khan could you update the description with some details on why you're making these changes around signals and how I can manually test everything here?

File: exams/utils.py (hunk 7 above)
Comment (rhysyngsun) [inline exams/utils.py]: I discussed with @pdpinch earlier and he agreed that we should only authorize once per course in this command and that the most recently passed one is probably the most reasonable one to use.

File: exams/signals.py (hunk 2 above)
Comment (amir-qayyum-khan) [inline exams/signals.py]: Done updated

File: exams/utils.py (hunk 7 above)
Comment (amir-qayyum-khan) [inline exams/utils.py]: So what if course is ongoing, should i look for status passed as well?

**Comment by amir-qayyum-khan** [comment]
@rhysyngsun removed signal code here

File: exams/management/commands/authorization_user_exam.py [hunk 9]
```diff
@@ -48,25 +37,9 @@ def handle(self, *args, **options):
...
+        bulk_authorize_for_exam(
+            program_id=program_id,
+            username=username
+        )
 
-                for enrollment in enrollments_qset:
-                    # if user has passed and paid for the course
-                    # and not already authorized for exam the create authorizations.
-                    authorize_for_exam(mmtrack, enrollment.course_run)
+        print("Done with authorization process")
```
Comment (giocalitri) [inline exams/management/commands/authorization_user_exam.py]: please do not use "print" statements: this is an example of how to proper write output of management commands (also with a nice output) taken from
https://docs.djangoproject.com/en/1.10/howto/custom-management-commands/
```
self.stdout.write(self.style.SUCCESS('Successfully closed poll "%s"' % poll_id))
```

File: exams/utils.py [hunk 10]
```diff
@@ -59,3 +65,59 @@ def authorize_for_exam(mmtrack, course_run):
...
+            course_ids = set()
+
+            for edx_course_key in mmtrack.edx_key_course_map.keys():
+                course_ids.add(mmtrack.edx_key_course_map[edx_course_key])
+
+            # get latest course_run from given course
+            for course_id in course_ids:
+                enrollments_qset = CachedEnrollment.user_course_qset(user, program=program).filter(
+                    course_run__course__id=course_id
+                ).order_by('course_run__end_date')
```
Comment (giocalitri) [inline exams/utils.py]: from what I read this should be ordered `desc`

File: exams/utils_test.py [hunk 11]
```diff
@@ -23,43 +27,49 @@
...
-            course_run=course_run,
-            data={
-                "passed": True,
-                "percent": 0.9,
-                "course_key": course_run.edx_course_key,
-                "username": cls.user.username
-            }
-        )
+        CachedEnrollmentFactory.create(user=cls.user, course_run=course_run)
+        with mute_signals(post_save):
```
Comment (giocalitri) [inline exams/utils_test.py]: why there is this `mute_signals`?

Comment (amir-qayyum-khan) [inline exams/utils_test.py]: it need to update changes status to in progress

File: exams/utils_test.py [hunk 12]
```diff
@@ -107,3 +117,111 @@ def test_exam_authorization_when_not_passed_course(self):
...
+    """Tests for authorization_user_exam command operations"""
+    @classmethod
+    def set_data(cls, user, program, course_run):
+        """
+        creates user records like CachedGrades, CachedEnrollments and CachedCertificate
+        """
+        with mute_signals(post_save):
+            ProgramEnrollment.objects.create(user=user, program=program)
+        CachedEnrollmentFactory.create(user=user, course_run=course_run)
+        with mute_signals(post_save):
```
Comment (giocalitri) [inline exams/utils_test.py]: why the `mute_signals` here?

File: exams/utils_test.py (hunk 11 above)
Comment (giocalitri) [inline exams/utils_test.py]: sorry I do not understand: what's the signal handler hooked to `CachedCurrentGrade`>

File: exams/utils_test.py [hunk 13]
```diff
@@ -107,3 +117,111 @@ def test_exam_authorization_when_not_passed_course(self):
...
+            profile1 = ProfileFactory.create()
+            profile2 = ProfileFactory.create()
+
+        cls.program, cls.tier_programs = create_program(
+            past=True,
+            exam_series_code="TestExam",
+            exam_module="SampleCourse"
+        )
+        cls.course_run = course_run = cls.program.course_set.first().courserun_set.first()
```
Comment (giocalitri) [inline exams/utils_test.py]: you do not need 2 variables here

File: exams/utils_test.py [hunk 14]
```diff
@@ -107,3 +117,111 @@ def test_exam_authorization_when_not_passed_course(self):
...
+        cls.users = users = [profile1.user, profile2.user]
```
Comment (giocalitri) [inline exams/utils_test.py]: and not even here you need 2 variables

File: financialaid/api_test.py [hunk 15]
```diff
@@ -64,7 +66,7 @@ def create_program(create_tiers=True, past=False):
         is_valid=True
     )
     tier_programs = None
-    if create_tiers:
+    if fa_availability and create_tiers:
```
Comment (giocalitri) [inline financialaid/api_test.py]: why this change?

**Comment by giocalitri** [comment]
In addition to the comments I already made, I think that `bulk_authorize_for_exam` should be split in multiple functions and each of them should address one step of `bulk_authorize_for_exam`,
but we can open a different issue for this.

**Comment by giocalitri** [comment]
I am doing functional testing and the command is ending (apparently with success) without doing anything (because of missing configurations): I do not think this should happen.
I am not sure if this should be part of this PR or not.

File: financialaid/api_test.py (hunk 15 above)
Comment (amir-qayyum-khan) [inline financialaid/api_test.py]: to create Non fa program

**Comment by giocalitri** [comment]
I also think that the management command by default should not do anything without parameters and if someone wants to run for everything, she should specify a `--all` argument, while without arguments the script should just fail and return the `--help`.
Or just make 2 different management commands, one for all the users and one for the specific program and/or user.
Not sure this belongs to this PR.

File: exams/utils.py [hunk 16]
```diff
@@ -59,3 +65,59 @@ def authorize_for_exam(mmtrack, course_run):
...
+            log.info('[authorization_user_exam] User(s) not exist for exam authorization.')
+            continue
+
+        for program_enrollment in program_enrollment_qset:
+            user = program_enrollment.user
+            mmtrack = get_mmtrack(user, program)
+            course_ids = set()
+
+            for edx_course_key in mmtrack.edx_key_course_map.keys():
+                course_ids.add(mmtrack.edx_key_course_map[edx_course_key])
```
Comment (giocalitri) [inline exams/utils.py]: cannot you just do
`course_ids = set(mmtrack.edx_key_course_map.values())`
or am I missing something?

Comment (giocalitri) [inline exams/utils.py]: I might be wrong, but I think you should filter also by courses that are finished. You might just filter by course_run that have a frozen final grade, (but this should not be done in this PR, probably)

Comment (amir-qayyum-khan) [inline exams/utils.py]: ok

**Comment by amir-qayyum-khan** [comment]
@giocalitri  i am logging success or failiure

**Comment by giocalitri** [comment]
`exam.utils.authorize_for_exam` needs to be refactored to raise exceptions in case of missing configurations.

**Comment by giocalitri** [comment]
> @giocalitri i am logging success or failiure

I am not sure which comment are you referring to, but if it is 

>I am doing functional testing and the command is ending (apparently with success) without doing anything (because of missing configurations): I do not think this should happen.
I am not sure if this should be part of this PR or not.

I think you are wrong: 
1. You use `log.info` that by default will be silenced (and it is silenced on my local env by default)
2. The code was stopping in my case inside `authorize_for_exam` because `ok_for_exam` was `False and there there is no logging of any kind.

**Comment by giocalitri** [comment]
Ok: I mocked my way through the code and I got the result.
I think that it will be @pdpinch decision if any of the suggestions above need to be implemented in this PR:
I advice against shipping without any kind of explicit error returned or raised by the management command.

File: exams/management/commands/authorization_user_exam.py [hunk 17]
```diff
@@ -32,41 +28,29 @@ def add_arguments(self, parser):
...
+            bulk_authorize_for_exam()
+            self.stdout.write(self.style.SUCCESS('Done with authorization process'))
+        elif program_id or username:
+            bulk_authorize_for_exam(
+                program_id=program_id,
+                username=username
+            )
+            self.stdout.write(self.style.SUCCESS('Done with authorization process'))
+        else:
+            self.stdout.write(self.style.SUCCESS('No param is provided.'))
```
Comment (giocalitri) [inline exams/management/commands/authorization_user_exam.py]: This should be modified to not return a success message (there is no success here): this should return the `--help` option.

Comment (amir-qayyum-khan) [inline exams/management/commands/authorization_user_exam.py]: done

File: exams/management/commands/authorization_user_exam.py [hunk 18]
```diff
@@ -32,41 +28,29 @@ def add_arguments(self, parser):
...
         )
 
     def handle(self, *args, **options):
         """
         Trigger exam authorizations for user for a course if he has passed and paid for it.
         """
-        # pylint: disable=too-many-locals
+        all = options.get("all", 'n')
         program_id = options.get("program_id", None)
         username = options.get("username", None)
```
Comment (giocalitri) [inline exams/management/commands/authorization_user_exam.py]: question: are you sure you need to specify the default if there is a `default=` option when you add the arguments before?

File: exams/management/commands/authorization_user_exam.py [hunk 19]
```diff
@@ -32,41 +28,29 @@ def add_arguments(self, parser):
...
-            "--course-key",
-            dest="course_key",
-            default=None,
-            help="Edx course key where we want to authorize users who passed course(s).",
-            type=int
+            "--all",
+            dest="all",
+            default='n',
+            help="Authorize eligible users for exams. Can be y or n",
+            type=str
```
Comment (giocalitri) [inline exams/management/commands/authorization_user_exam.py]: this should just be an argument without value, meaning I should be able to specify `--all` and not `--all y`

File: exams/management/commands/authorization_user_exam.py (hunk 18 above)
Comment (amir-qayyum-khan) [inline exams/management/commands/authorization_user_exam.py]: yes, you are right, we dont need

File: exams/management/commands/authorization_user_exam.py (hunk 19 above)
Comment (amir-qayyum-khan) [inline exams/management/commands/authorization_user_exam.py]: Done

File: exams/signals_test.py [hunk 20]
```diff
@@ -38,7 +38,11 @@ def setUpTestData(cls):
         with mute_signals(post_save):
             cls.profile = profile = ProfileFactory.create()
 
-        program, _ = create_program(past=True)
+        program, _ = create_program(
+            past=True,
+            exam_series_code="TestExam",
+            exam_module="SampleCourse"
+        )
```
Comment (giocalitri) [inline exams/signals_test.py]: is this a leftover of the PR split or do you need it?

File: exams/utils.py [hunk 21]
```diff
@@ -59,3 +65,84 @@ def authorize_for_exam(mmtrack, course_run):
...
+    """
+    Authorize user(s) for exam(s).
+
+    Args:
+        program_id(str): Program id (optional) if you want authorization on specific program exams
+        username(str): User name (optional) whom you want to authorize for exam(s)
+    """
+    programs = Program.objects.filter(
+        live=True
+    ).exclude(exam_series_code__isnull=True).exclude(exam_series_code__exact="")
```
Comment (giocalitri) [inline exams/utils.py]: nit: you can use the `Q` operator here and use only one `.exclude`

**Comment by giocalitri** [comment]
you need to write a test for the management command just to assert that the right function with the right parameters is called.

**Comment by giocalitri** [comment]
@pdpinch please test this and see if the behavior (erroneous cases included) is good enough.

This need to be refactored anyway to remove the logging and raise exceptions to be explicit about where the code stopped, but we can open an issue about this.

File: exams/signals_test.py (hunk 20 above)
Comment (amir-qayyum-khan) [inline exams/signals_test.py]: nop, these conditions must meet for user to be authorize for exam.

Comment (giocalitri) [inline exams/signals_test.py]: I understand, but this file is `signal_test` and you have not modified the file `signal`, so why there is a change here?

File: dashboard/utils.py [hunk 22]
```diff
@@ -371,6 +370,18 @@ def get_pearson_exam_status(self):  # pylint: disable=too-many-return-statements
...
+        """
+        Check if the user is authorized for an exam for a course run
+        """
+        return bool(
+            course_run and
+            course_run.edx_course_key and
+            course_run.course and
+            course_run.course.exam_module and
+            self.program.exam_series_code
+        )
```
Comment (giocalitri) [inline dashboard/utils.py]: why this is here and it is not a property of the `CourseRun` model?

File: exams/signals_test.py (hunk 20 above)
Comment (giocalitri) [inline exams/signals_test.py]: you changed `exams/signals.py` but there are no new tests for the changes

File: exams/utils.py [hunk 23]
```diff
@@ -59,7 +68,105 @@ def authorize_for_exam(mmtrack, course_run):
...
+                user=mmtrack.user.username,
+                course_id=course_run.edx_course_key
+            )
+            raise ExamAuthorizationException(errors_message)
+    else:
+        errors_message = message_not_eligible_template.format(
+            user=mmtrack.user.username,
+            course_id=course_run.edx_course_key
+        )
+        raise ExamAuthorizationException(errors_message)
```
Comment (giocalitri) [inline exams/utils.py]: please refactor this code in the following way (please verify that it actually works):
```
def authorize_for_exam(mmtrack, course_run):
    """
    Authorize user for exam if he has paid for course and passed course.
    Args:
        mmtrack (dashboard.utils.MMTrack): a instance of all user information about a program.
        course_run (CourseRun): A CourseRun object.
    """
    if not mmtrack.course_has_exam(course_run) and mmtrack.has_paid(course_run.edx_course_key):
        raise ExamAuthorizationException(
            message_not_eligible_template.format(
                user=mmtrack.user.username,
                course_id=course_run.edx_course_key
            )
        )

    now = datetime.datetime.now(tz=pytz.UTC)
    # if user paid for a course then create his exam profile if it is not creaated yet.
    ExamProfile.objects.get_or_create(profile=mmtrack.user.profile)

    # if user passed the course and currently not authorization for that run then give
    # her authorizations.
    ok_for_authorization = (
        mmtrack.has_passed_course(course_run.edx_course_key) and
        not ExamAuthorization.objects.filter(
            user=mmtrack.user,
            course=course_run.course,
            date_first_eligible__lte=now,
            date_last_eligible__gte=now
        ).exists()
    )

    if not ok_for_authorization:
        raise ExamAuthorizationException(
            message_not_passed_or_exist_template.format(
                user=mmtrack.user.username,
                course_id=course_run.edx_course_key
            )
        )

    ExamAuthorization.objects.create(
        user=mmtrack.user,
        course=course_run.course,
        date_first_eligible=now,
        date_last_eligible=add_year(now)
    )
    log.info(
        '[Exam authorization] user "%s" is authorize for exam the for course id "%s"',
        mmtrack.user.username,
        course_run.edx_course_key
    )
```

File: exams/utils.py [hunk 24]
```diff
@@ -59,7 +68,105 @@ def authorize_for_exam(mmtrack, course_run):
...
+        exam_series_code__isnull=False
+    )
+    if program_id:
+        programs = programs.filter(id=program_id)
+
+    if not programs.exists():
+        if program_id:
+            raise CommandError('[Exam authorization] Invalid program id: %s', program_id)
+        else:
+            raise CommandError('[Exam authorization] Program(s) are not available for exam authorization.')
```
Comment (giocalitri) [inline exams/utils.py]: `CommandError` is an exception to be used in the management commands, not here.

File: exams/utils.py [hunk 25]
```diff
@@ -59,7 +68,105 @@ def authorize_for_exam(mmtrack, course_run):
...
+    Args:
+        program(courses.models.Program): Program object.
+        username(str): User name (optional) whom you want to authorize for exam(s)
+    """
+    program_enrollment_qset = ProgramEnrollment.objects.filter(program=program)
+    if username:
+        program_enrollment_qset = program_enrollment_qset.filter(user__username=username)
+
+    if not program_enrollment_qset.exists():
+        if username:
```
Comment (giocalitri) [inline exams/utils.py]: please be explicit
`if username is not None`

File: exams/utils.py [hunk 26]
```diff
@@ -59,7 +68,105 @@ def authorize_for_exam(mmtrack, course_run):
...
+
+def authorize_for_exam_given_program(program, username=None):
+    """
+    Authorize user(s) for exam on given program.
+    Args:
+        program(courses.models.Program): Program object.
+        username(str): User name (optional) whom you want to authorize for exam(s)
+    """
+    program_enrollment_qset = ProgramEnrollment.objects.filter(program=program)
+    if username:
```
Comment (giocalitri) [inline exams/utils.py]: please be explicit
`if username is not None`

File: exams/utils.py [hunk 27]
```diff
@@ -59,7 +68,105 @@ def authorize_for_exam(mmtrack, course_run):
...
+        program_enrollment_qset = program_enrollment_qset.filter(user__username=username)
+
+    if not program_enrollment_qset.exists():
+        if username:
+            raise CommandError('[Exam authorization] Invalid username: %s', username)
+        else:
+            raise CommandError('[Exam authorization] Eligible users do not exist for exam authorization.')
```
Comment (giocalitri) [inline exams/utils.py]: `CommandError` is an exception to be used in the management commands, not here.

File: exams/management/commands/test_authorization_user_exam.py [hunk 28]
```diff
@@ -0,0 +1,148 @@
...
+        with self.assertRaises(CommandError) as ce:
+            self.command.handle("authorization_user_exam", username="invalid_user")
+
+        assert ce.exception.args[0] == '[Exam authorization] Invalid username: %s', "invalid_user"
+        for user in self.users:
+            self.assertFalse(ExamProfile.objects.filter(profile=user.profile).exists())
+            self.assertFalse(ExamAuthorization.objects.filter(
+                user=user,
+                course=self.course_run.course
+            ).exists())
```
Comment (giocalitri) [inline exams/management/commands/test_authorization_user_exam.py]: I think that these tests are not necessary: the only thing to be done to test the management command is to assert that when the command is called, the right backend function is called.
So you can mock the entire backend function and just assume that it works.

File: dashboard/utils.py (hunk 22 above)
Comment (amir-qayyum-khan) [inline dashboard/utils.py]: yes it is not part of `CourseRun`, but for course to have exam its program must have  exam_series_code.

File: exams/signals_test.py (hunk 20 above)
Comment (amir-qayyum-khan) [inline exams/signals_test.py]: this change is because signal uses `exams.utils.authorize_for_exam` and we change few things there

**Comment by amir-qayyum-khan** [comment]
Done @giocalitri

File: exams/management/commands/test_authorization_user_exam.py [hunk 29]
```diff
@@ -0,0 +1,86 @@
...
+        cls.program, _ = create_program(past=True)
+        cls.course_run = cls.program.course_set.first().courserun_set.first()
+        cls.users = [profile1.user, profile2.user]
+
+        for user in cls.users:
+            with mute_signals(post_save):
+                ProgramEnrollment.objects.create(user=user, program=cls.program)
+            cls.set_data(user, cls.course_run)
+
+        cls.command = authorization_user_exam.Command()
```
Comment (giocalitri) [inline exams/management/commands/test_authorization_user_exam.py]: do you still need all this data in the set up?

File: dashboard/utils.py [hunk 30]
```diff
@@ -372,6 +371,18 @@ def get_pearson_exam_status(self):  # pylint: disable=too-many-return-statements
...
```
Comment (amir-qayyum-khan) [inline dashboard/utils.py]: just to check if course has exam we need bool statement that will give us true or false

File: dashboard/utils.py (hunk 22 above)
Comment (amir-qayyum-khan) [inline dashboard/utils.py]: you are right @giocalitri , i just moved it to course_run

File: dashboard/utils.py (hunk 30 above)
Comment (amir-qayyum-khan) [inline dashboard/utils.py]: Fixed there was issue in logic

File: courses/utils.py [hunk 31]
```diff
@@ -89,3 +89,16 @@ def get_year_season_from_course_run(course_run):
         return season_year_tuple[0], NUMBER_SEASON_MAP[season_year_tuple[1]]
     else:
         return season_year_tuple
+
+
+def is_blank(string):
```
Comment (giocalitri) [inline courses/utils.py]: please do not use reserved keywords (`string` is a module of the standard library)

File: courses/utils_test.py [hunk 32]
```diff
@@ -34,3 +42,13 @@ def test_get_year_season_from_course_run(self):
...
         assert get_year_season_from_course_run(unparseable_test_run1) == ()
         assert get_year_season_from_course_run(unparseable_test_run2) == ()
+
+    @data(
+        (None, True),
+        ("", True),
+        ("Foo", False),
+    )
+    @unpack
+    def test_is_blank(self, string, expected):
```
Comment (giocalitri) [inline courses/utils_test.py]: please do not use reserved keywords

File: dashboard/utils.py [hunk 33]
```diff
@@ -362,7 +361,7 @@ def get_pearson_exam_status(self):  # pylint: disable=too-many-return-statements
             course__program=self.program
         )
 
-        if not any(course_has_exam(self, cr) for cr in course_runs_for_program):
+        if not any(cr.has_exam for cr in course_runs_for_program):
```
Comment (giocalitri) [inline dashboard/utils.py]: nit: use a more readable variable

**Comment by amir-qayyum-khan** [comment]
Done @giocalitri

**Comment by giocalitri** [comment]
Can you please rebase master?

**Comment by pdpinch** [comment]
@amir-qayyum-khan can you update the name (and maybe the description) of this to better describe what's covered by this?

Also, it looks like this needs a rebase

**Comment by giocalitri** [comment]
The code looks fine: doing a manual test

File: exams/utils.py [hunk 34]
```diff
@@ -79,44 +79,151 @@ def authorize_for_exam(mmtrack, course_run):
...
+        ~Q(exam_series_code__exact=""),
+        live=True,
+        exam_series_code__isnull=False
+    )
+    if program_id is not None:
+        programs = programs.filter(id=program_id)
+
+    if not programs.exists():
+        if program_id is not None:
+            raise ExamAuthorizationException('[Exam authorization] Invalid program id: %s', program_id)
```
Comment (giocalitri) [inline exams/utils.py]: I did a test without setting the `program.exam_series_code` and the management command correctly raises an exception at this point, but the message is confusing because the program is right, it is the `exam_series_code` that is missing.
Can you please update this error message saying that either the `program id` is wrong or the `exam_series_code` is missing?

**Comment by giocalitri** [comment]
If I do not set a `course.exam_module` I do not get any error and the command seems to run fine but I have nothing in `ExamAuthorization`

File: exams/utils.py (hunk 34 above)
Comment (amir-qayyum-khan) [inline exams/utils.py]: or should i make it, `program: 'x' is not eligible for exam`

Comment (giocalitri) [inline exams/utils.py]: better to be explicit in this case because `program: 'x' is not eligible for exam` is still ambiguous

**Comment by giocalitri** [comment]
ok: functionality looks decent at the moment, but I think we should improve the tool to list all the users that have been authorized to have a better understanding of what's going on.
Please open an issue for this.

**Comment by amir-qayyum-khan** [comment]
Done @giocalitri https://github.com/mitodl/micromasters/issues/2535

**Comment by giocalitri** [comment]
what about the message and the other problem?

**Comment by amir-qayyum-khan** [comment]
i am working on those, thats why i did not change the PR status

**Comment by amir-qayyum-khan** [comment]
Done

**Comment by giocalitri** [comment]
👍

**Comment by giocalitri** [comment]
please squash
